driver/at86rf2xx: optimized auto initialization

haukepetersen opened the pull request:
simplified the auto configuration for the Atmel radio driver by introducing a default `at86rf2xx_params.h` file.

This PR is meant as an example how I could imagine all devices drivers in RIOT to be structured in the future... (see #4271 for a description of the proposed concept)

OlegHahm commented:
I like.

authmillenon commented:
Me too :-)

cgundogan commented:
Seems reasonable. Reduces a lot of code duplication.

haukepetersen commented:
so anyone bold enough to ack this? @kaspar030?

kaspar030 commented:
I like a lot that at86rf2xx_params.h exists only once.

Still I'd prefer having the configuration in C structs. The whole point of introducing *_param_t style structs was to end up with iterable, type checked configuration data instead of error-prone and clumsy defines.

What do you thing about adopting the shared at86rf_params_t from this PR, but put something like this in ```periph_conf.h``` (per board):

```
#define HAS_ONBOARD_AT86RF2XX

static const at86rf2xx_params_t _at86rf2xx_onboard = {
            .spi = SPI0,
            .spi_speed = blah,
            .cs_pin = blubb,
            .int_pin = pin,
            .sleep_pin = xx,
            .reset_pin = xx,
        };
```
(not an array, just one at86rf2xx_params_t)

and conditionally add that to the default ```at86rf2xx_params.h``` array?
Then later we can also include e.g., a "periph_conf_application.h" and add more conditional entries later.

haukepetersen commented:
changed the default config file slightly again - the default board configuration is now given as a single define. This has the advantage, that it can be simply re-used in any overriding `xx_params.h` file...

@kaspar030: is this what you had in mind?

kaspar030 commented:
Yes!

If you change the definitions in board.h to { .spi=blah, .spi_speed=blah ...}, the compiler has even more information to detect errors.

haukepetersen commented:
true, also nicer to read. Will do

haukepetersen commented:
done.

kaspar030 commented on drivers/at86rf2xx/include/at86rf2xx_params.h:
```diff
@@ -0,0 +1,78 @@
+#define AT86RF2XX_PARAM_RESET       (GPIO_PIN(0, 3))
+#endif
+
+#define AT86RF2XX_PARAMS_DEFAULT    {AT86RF2XX_PARAM_SPI, \
```
you missed these... :)

haukepetersen replied:
ups

kaspar030 commented:
I like, please squash!

haukepetersen commented:
fixed and squashed.

OlegHahm commented:
Ready for merging? @kaspar030 you acked, right?

haukepetersen commented:
not explicitly :-)

kaspar030 commented:
I only have one board with the chip, so I can't actually test the radio... ACK if someone can confirm it's still working.

OlegHahm commented:
Hm, seems to break something on iotlab-m3. `ifconfig` doesn't show any output any more.

haukepetersen commented:
ups, breaking through rebasing... I'll look into it

haukepetersen commented:
fixed (and squashed): the sleep and reset pins were switched around for the `fox` and the `iotlab`. Also fixed a missing newline in the auto-init/at86rf2xx file..

OlegHahm commented:
Looks good. ACK and go as soon as Travis agrees.
